File: Java/백준/Gold/10830. 행렬 제곱/행렬 제곱.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
    static int N;
    static long B;
    static int[][] A;

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        B = Long.parseLong(st.nextToken());
        A = new int[N][N];
        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < N; j++) {
                A[i][j] = Integer.parseInt(st.nextToken()) % 1000;
            }
        }
        int[][] result = pow(A, B);
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < N; j++) {
                sb.append(result[i][j] % 1000).append(" ");
            }
            sb.append("\n");
        }
        System.out.print(sb);
    }

    static int[][] pow(int[][] matrix, long exp) {
        if (exp == 1) return matrix;
        int[][] half = pow(matrix, exp / 2);
        int[][] temp = multiply(half, half);
        if (exp % 2 == 0) return temp;
        else return multiply(temp, A);
    }

    static int[][] multiply(int[][] m1, int[][] m2) {
        int[][] res = new int[N][N];
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < N; j++) {
                long sum = 0;
                for (int k = 0; k < N; k++) {
                    sum += (m1[i][k] * m2[k][j]);
                }
                res[i][j] = (int)(sum % 1000);
            }
        }
        return res;
    }
}

File: C11/백준/Gold/10830. 행렬 제곱/행렬 제곱.c
#include <stdio.h>

#define MOD 1000
#define MAX 5

int N;
long long B;
int A[MAX][MAX];

void multiply(int a[MAX][MAX], int b[MAX][MAX], int res[MAX][MAX]) {
    int temp[MAX][MAX] = {0};
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) {
            long long sum = 0;
            for (int k = 0; k < N; k++) {
                sum += (long long)a[i][k] * b[k][j];
            }
            temp[i][j] = sum % MOD;
        }
    }
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) {
            res[i][j] = temp[i][j];
        }
    }
}

void power(int base[MAX][MAX], long long exp, int result[MAX][MAX]) {
    if (exp == 1) {
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < N; j++) {
                result[i][j] = base[i][j] % MOD;
            }
        }
        return;
    }

    int half[MAX][MAX];
    power(base, exp / 2, half);

    int temp[MAX][MAX];
    multiply(half, half, temp);

    if (exp % 2 == 0) {
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < N; j++) {
                result[i][j] = temp[i][j];
            }
        }
    } else {
        multiply(temp, A, result);
    }
}

int main() {
    scanf("%d %lld", &N, &B);
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) {
            scanf("%d", &A[i][j]);
            A[i][j] %= MOD;
        }
    }

    int result[MAX][MAX];
    power(A, B, result);

    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) {
            printf("%d ", result[i][j] % MOD);
        }
        printf("\n");
    }
    return 0;
}